File: 39. Combination Sum/_39.py
class Solution:
    # dfs
    def combinationSum(self, candidates, target: int):
        candidates.sort()
        res = []
        self.findSum(candidates, target, res, [])
        return res
        
    def findSum(self, candidates, target, res, tmp):
        if target<0:
            return
        if target == 0:
            res.append(tmp)
            return
        for i in range(len(candidates)):
            if target-candidates[i] < 0:
                break
            self.findSum(candidates[i:], target-candidates[i], res, tmp+[candidates[i]])

    # dp
    def combinationSum1(self, nums, target: int) -> int:
        # 保存方案，使用三维数组
        # 第一维： dp:List 表示和为不同数时的方案
        # 第二维： dp[s]:List 表示和i的方案列表
        # 第三维 每种具体的方案都是数组， dp[s][j]:List 表示和为s的第j个方案的数组
        dp: list = [[[]]] + [[] for i in range(target)]  # 初始化 eg: [[[]], [], [], [], [], []]， 保证和为0有一种方案 [], 完全背包需要恰好装满的时候，dp[0] 必须初始化为0

        for n in nums:  # 对于每个数
            for s in range(n, target + 1):  # 完全背包，从小到大
                li = [l + [n] for l in dp[s - n]]  # 对于把和为s-n的每种方案，添加数字n, 就变成一种新的 和为s的方案
                dp[s] += li  # 把新方案添加到 原来和为s的方案里
        return dp[target]

# a = Solution()
# print(a.combinationSum([2,3,5], 8))

File: 39. Combination Sum/39.go
package leetcode

import(
	"sort"
)

func combinationSum(candidates []int, target int) [][]int {
	res := make([][]int, 0)
	sort.Ints(candidates)
	dfs(candidates, target, &res, []int{})
	return res
}

func dfs(candidates []int, target int, res *[][]int, tmp []int){
	if target<0{
		return
	}
	if target == 0{
		*res = append(*res, tmp)
		return
	}
	for i, v := range(candidates){
		if v>target{
			return
		}
		tmp1 := make([]int, 0, len(tmp))
		tmp1 = append(tmp1, tmp...)
		tmp1 = append(tmp1, v)
		dfs(candidates[i:], target-v, res, tmp1)
	}
}

func combinationSum1(candidates []int, target int) [][]int {//最简单的想法，直接回溯剪枝
    sort.Ints(candidates)//排个序就能节省大量的去重过程
    var results [][]int
    var result []int
    var sum int
    find(candidates,target,&result,&results,sum)
    return results
}

func find(candidates []int,target int,result *[]int,results *[][]int,sum int){
    for i:=0;i<len(candidates);i++{
        if sum+candidates[i]<target{//继续往下找
            *result=append(*result,candidates[i])
            find(candidates[i:],target,result,results,sum+candidates[i])
            *result=(*result)[:len(*result)-1]
        }else{//找到或者已经超出，结束这个分支
            if sum+candidates[i]==target{//找到
                *result=append(*result,candidates[i])
                var cc []int
                for _,v:=range *result{//不能直接用从*result切出来的，因为那是地址引用
                    cc=append(cc,v)
                }
                *results=append(*results,cc)
                *result=(*result)[:len(*result)-1]//让*result回溯之前返回原样
            }
            continue//这个分支结束了
        }
    }
}